"""empty message

Revision ID: 9f72c83a95f5
Revises: c7db23d26c06
Create Date: 2023-01-10 14:34:44.532014

"""
from alembic import op
import sqlalchemy as sa


# revision identifiers, used by Alembic.
revision = '9f72c83a95f5'
down_revision = 'c7db23d26c06'
branch_labels = None
depends_on = None


def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('amr_finder_result',
    sa.Column('id', sa.Integer(), nullable=False),
    sa.Column('contig_id', sa.VARCHAR(length=60), nullable=True, comment='The contig the gene is located on'),
    sa.Column('start', sa.Numeric(), nullable=True, comment='The start position of the gene on the genome'),
    sa.Column('stop', sa.Numeric(), nullable=True, comment='The stop position of the gene on the genome'),
    sa.Column('strand', sa.VARCHAR(length=60), nullable=True, comment='The strand the gene is located on'),
    sa.Column('gene_symbol', sa.VARCHAR(length=60), nullable=True, comment='The symbol of gene found'),
    sa.Column('sequence_name', sa.VARCHAR(length=60), nullable=True, comment='The sequence name of the gene found'),
    sa.Column('gene_class', sa.VARCHAR(length=60), nullable=True, comment='The sequence class of the gene found'),
    sa.Column('gene_subclass', sa.VARCHAR(length=60), nullable=True, comment='The sequence subclass of the gene found'),
    sa.Column('target_length', sa.Numeric(), nullable=True, comment='The gene length on the genome'),
    sa.Column('reference_sequence_length', sa.Numeric(), nullable=True, comment='The reference gene length'),
    sa.Column('pct_cov_of_ref_seq', sa.Numeric(), nullable=True, comment='Percentage coverage of reference sequence'),
    sa.Column('pct_id_to_ref_seq', sa.Numeric(), nullable=True, comment='Percentage identity of reference sequence'),
    sa.Column('alignment_length', sa.Numeric(), nullable=True, comment='The alignment length of the target to the reference'),
    sa.Column('accession_of_closest_sequence', sa.VARCHAR(length=60), nullable=True, comment='The sequence subclass of the gene found'),
    sa.Column('amr_version', sa.VARCHAR(length=60), nullable=True, comment='The AMRFinder version used to collect the results'),
    sa.PrimaryKeyConstraint('id')
    )
    op.create_table('assembly',
    sa.Column('id', sa.Integer(), nullable=False),
    sa.Column('assembler', sa.VARCHAR(length=60), nullable=True, comment='The assembler software used to generate draft genome.'),
    sa.Column('version', sa.VARCHAR(length=60), nullable=True, comment='The version of the assembler software.'),
    sa.Column('n50', sa.Numeric(), nullable=True, comment='The N50 of the draft genome.'),
    sa.Column('num_of_contigs', sa.Numeric(), nullable=True, comment='The number of contigs generated by assembly.'),
    sa.Column('total_assembled_length', sa.Numeric(), nullable=True, comment='The length of draft genome.'),
    sa.PrimaryKeyConstraint('id')
    )
    # ### end Alembic commands ###


def downgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.drop_table('assembly')
    op.drop_table('amr_finder_result')
    # ### end Alembic commands ###
